[CARBONDATA-1827] S3 Carbon Implementation

jatin9896 opened the pull request:
1) Provide support for s3 in carbondata.
2) Added S3Example to create store on s3.
3) Added S3CSVExample to load csv from s3.
 

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed? NO
 
 - [x] Any backward compatibility impacted? NO
 
 - [x] Document update required? NO

 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required? 
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       Added Examples to test the functionality
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1566/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2801/

anubhav100 commented:
retest this please

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2896/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1571/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2807/

anubhav100 commented:
retest this please

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2899/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1572/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2808/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2810/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1574/

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,151 @@
+    import org.apache.spark.sql.CarbonSession._
+    if (args.length != 4) {
+      logger.error("Usage: java CarbonS3Example <access-key> <secret-key>" +
+                   "<carbon store location> <spark-master>")
```
arg(2) is table path, not store path, right?

jatin9896 replied:
you are right. ok I will change it to tablepath.

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,151 @@
+   * @param args require three parameters "Access-key" "Secret-key"
+   *             "s3 bucket path" "spark-master"
+   */
+
```
remove empty line

jatin9896 replied:
removed.

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,151 @@
+object S3Example {
+
+  /**
+   * This example demonstrate usage of s3 as a store.
```
Change to 
> This example demonstrate usage of
> 1. create carbon table with storage location on object based storage like AWS S3, Huawei OBS, etc
> 2. load data into carbon table, the generated file will be stored on object based storage
> 3. query the table.
> With the indexing feature of carbondata, the data read from object based storage is minimum, thus providing both high performance analytic and low cost storage

jatin9896 replied:
done

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,151 @@
+      .addProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT, "yyyy/MM/dd HH:mm:ss")
+      .addProperty(CarbonCommonConstants.CARBON_DATE_FORMAT, "yyyy/MM/dd")
+      .addProperty(CarbonCommonConstants.ENABLE_UNSAFE_COLUMN_PAGE_LOADING, "true")
+      .addProperty(CarbonCommonConstants.DEFAULT_CARBON_MAJOR_COMPACTION_SIZE, "0.02")
```
remove unnecssary property setting, all these are not required

jatin9896 replied:
done

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,151 @@
+         | OPTIONS('HEADER'='true')
+       """.stripMargin)
+
+    spark.sql("ALTER table carbon_table compact 'MINOR'")
```
Add a comment: 
> // Use compaction command to merge segments or small files in object based storage, this can be done periodically.

jatin9896 replied:
Added

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala:
```diff
@@ -0,0 +1,111 @@
+    CarbonProperties.getInstance()
+      .addProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT, "yyyy/MM/dd HH:mm:ss")
+      .addProperty(CarbonCommonConstants.CARBON_DATE_FORMAT, "yyyy/MM/dd")
+      .addProperty(CarbonCommonConstants.ENABLE_UNSAFE_COLUMN_PAGE_LOADING, "true")
```
remove unnecssary property setting, all these are not required

jatin9896 replied:
Removed all properties.

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,151 @@
+    spark.sql("ALTER table carbon_table compact 'MINOR'")
+    spark.sql("show segments for table carbon_table").show()
+
+    spark.sql(
```
remove this to make it simpler

jackylk commented on core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java:
```diff
@@ -168,6 +168,14 @@
 
   public static final String S3A_PREFIX = "s3a://";
 
+  public static final String S3N_ACCESS_KEY = "fs.s3n.awsAccessKeyId";
```
Add comment for these 4 properties

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala:
```diff
@@ -0,0 +1,111 @@
+   * @param args require three parameters "Access-key" "Secret-key"
+   *             "s3 path to csv" "spark-master"
+   */
+
```
remove empty line

jatin9896 replied:
done

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2906/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1585/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2820/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1598/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2916/

jackylk commented:
retest this please

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1615/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2851/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1621/

anubhav100 commented:
retest this please

jackylk commented:
Can you show an example of how to give <table_path> in S3Example using AWS S3? I tried with Huawei OBS, it has two problems:
1. I need to set the endpoint conf in main function by `spark.sparkContext.hadoopConfiguration.set("fs.s3a.endpoint", "obs.cn-north-1.myhwclouds.com")
` manually
2. After I set the conf, there is an exception thrown when running it: 
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/http/pool/ConnPoolControl

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2936/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2863/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1630/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2882/

jatin9896 commented:
@jackylk table path is the path to bucket location like I have provided s3a://<bucket-name>/<location> and regarding endpoints, I have modified example which takes endpoint as args(4) and it is not mandatory to provide. About connection pooling exception in the example is also fixed. Please check.

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1650/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2892/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1661/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2895/

anubhav100 commented:
retest this please

jackylk commented:
I tried, it is successful now. But in log it is showing using LOCAL_SORT, it should be NO_SORT, right?

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,157 @@
+      .config(endpoint, getS3EndPoint(args))
+      .getOrCreateCarbonSession()
+
+    spark.sparkContext.setLogLevel("INFO")
```
change to WARN, it is printing too many

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,157 @@
+         | OPTIONS('HEADER'='true')
+       """.stripMargin)
+
+    spark.sql(
```
do a select * after 1 load

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,157 @@
+         | INTO TABLE carbon_table
+         | OPTIONS('HEADER'='true')
+       """.stripMargin)
+
```
Now we do not allow doing compaction and data load concurrently, it will raise exception: Cannot run data loading and compaction on same table concurrently. Please wait for load to finish

So, I think you should check the finish of data load then do compaction, you can check finish of data load by SHOW SEGMENT and collect the result, loop until there are 3 segments

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,157 @@
+
+    // Use compaction command to merge segments or small files in object based storage,
+    // this can be done periodically.
+    spark.sql("ALTER table carbon_table compact 'MINOR'")
```
change this to MAJOR, and remove subsequent command, to make it simpler

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,157 @@
+         | OPTIONS('HEADER'='true')
+       """.stripMargin)
+
+    spark.sql(
```
remove one load to make it simpler

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,157 @@
+   * 1. create carbon table with storage location on object based storage
+   *    like AWS S3, Huawei OBS, etc
+   * 2. load data into carbon table, the generated file will be stored on object based storage
+   *    query the table.
```
remove this line

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,157 @@
+   *    thus providing both high performance analytic and low cost storage
+   *
+   * @param args require three parameters "Access-key" "Secret-key"
+   *             "s3 bucket path" "spark-master" "s3-endpoint"
```
modify to  `table path on S3`

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,157 @@
+    import org.apache.spark.sql.CarbonSession._
+    if (args.length < 4 || args.length > 5) {
+      logger.error("Usage: java CarbonS3Example <access-key> <secret-key>" +
+                   "<table-path> <spark-master> <s3-endpoint>")
```
Modify to `<table-path-on-S3>`

jackylk replied:
move <spark-master> as the last parameter, it should be optional, default value is 'local'

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1679/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2912/

jackylk commented:
please rebase and drop the first two commits which are merged to carbonstore branch already

ravipesala commented:
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2961/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1703/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2937/

jackylk commented:
LGTM

jatin9896 commented:
retest this please

jackylk commented on integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/AlterTableLoadPartitionRDD.scala:
```diff
@@ -41,102 +41,95 @@ class AlterTableLoadPartitionRDD[K, V](alterPartitionModel: AlterPartitionModel,
-    val databaseName = carbonTable.getDatabaseName
-    val factTableName = carbonTable.getTableName
-    val partitionInfo = carbonTable.getPartitionInfo(factTableName)
+  var storeLocation: String = null
```
Can you put these codestyle modification into a separate PR?

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,173 @@
+    // this can be done periodically.
+    spark.sql("ALTER table carbon_table compact 'MAJOR'")
+    spark.sql("show segments for table carbon_table").show()
+
```
please do one more select * query

jatin9896 replied:
added one more select query.

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  def getS3EndPoint(args: Array[String]): String = {
+    if (args.length >= 4 && args(3).contains(".com")) {
```
why need to compare with ".com"?

jatin9896 replied:
it might happen that at args(4) be spark master.

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,173 @@
+    if (args.length >= 4 && args(3).contains(".com")) {
+      args(3)
+    }
+    else {
```
move one line up

jatin9896 replied:
done

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,173 @@
+
+  def getSparkMaster(args: Array[String]): String = {
+    if (args.length >= 4) {
+      if (args.length == 5) {
```
this logic can be optimized, you can make one level of `if... else if`

jatin9896 replied:
sure

jatin9896 replied:
done

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,173 @@
+      if (args.length == 5) {
+        args(4)
+      }
+      else if (args(3).contains("spark:") || args(3).contains("mesos:")) {
```
`else if` should be same line as `}`

jatin9896 replied:
updated to same line.

jackylk commented on examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala:
```diff
@@ -0,0 +1,173 @@
+        "local"
+      }
+    }
+    else {
```
`else if` should be same line as `}`

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1714/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2948/

jackylk commented:
LGTM

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2953/

jackylk commented:
It is merged, please close it manually

ravipesala commented:
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2978/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1724/
